File: content/problems/1155-number-of-dice-rolls-with-target-sum/solutions/dice-count-rolling-dp/solution.cpp
class Solution {
    public: int numRollsToTarget(int n,int k,int target){
        const int MOD=1000000007;
        vector<long long> dp(target+1);
        dp[0]=1;
        while(n--){
            vector<long long> next(target+1);
            for(int sum=1;sum<=target;++sum)for(int face=1;face<=k&&face<=sum;++face)next[sum]=(next[sum]+dp[sum-face])%MOD;
            dp.swap(next);
        }
        return dp[target];
    }
};

File: content/problems/1155-number-of-dice-rolls-with-target-sum/solutions/dice-count-rolling-dp/Solution.kt
class Solution {
    fun numRollsToTarget(n:Int,k:Int,target:Int):Int{
        val mod=1000000007L;
        var dp=LongArray(target+1);
        dp[0]=1;
        repeat(n){
            val next=LongArray(target+1);
            for(sum in 1..target)for(face in 1..minOf(k,sum))next[sum]=(next[sum]+dp[sum-face])%mod;
            dp=next
        };
        return dp[target].toInt()
    }
}
